### PR TITLE
feat(performance-metrics): run system resource tracker as systemd service

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Overview

Add the definition of systemd service files and scripts to ensure that they work

Closes [EXEC-596](https://opentrons.atlassian.net/browse/EXEC-596)
 
# Changelog

- Defined 2 different service files a prod and dev
- Added systemd-python so systemd can be notified when service is ready
- Added some helper scripts for setting up and cleaning up on remote server

# Review requests

Take a look at my service files. This is my first time creating them. 
I read docs on how to do it and they work but would like some special attention paid to them.

# Risk assessment

Still low, nothing is installed or running by default. You have to run scripts to get them on the robot

# Test Plan

**Putting this at the end since it's long**

- [ ] pull this branch
- [ ] Go into `performance-metrics` directory
- [ ] run `make teardown && make setup` as dependencies have changed
- [ ] Run `make setup-performance-metrics-project-on-remote-flex host=<robot-ip> ssh_key=<key-path>`
- [ ] ssh to robot 
    - [ ] Run `cat /data/feature_flags.json` and ensure `enabledPerformanceMetrics` is set to `true`
    - [ ] exit ssh
- [ ] Run `make cleanup-remote-server host=<robot-ip> ssh_key=<key-path>`
  - [ ] As long as you haven't already configured the service the output should look like this:
```
Cleaning up performance metrics on host 10.10.10.151...
Configuring remote file system to read/write
Stopping system-resource-tracker
Failed to stop system-resource-tracker.service: Unit system-resource-tracker.service not loaded.
Disabling system-resource-tracker
Failed to disable unit: Unit file system-resource-tracker.service does not exist.
Reloading daemon
Checking system-resource-tracker status
inactive
Removing service file, data folders, and performance_metrics package
Remounting filesystem as read-only
Unsetting performance metrics feature flag
```
- [ ] Run `make setup-performance-metrics-project-on-remote-flex host=<robot-ip> ssh_key=<key-path>` as the clean command removed and unset everything
- [ ] Run `make setup-prod-system-resource-tracker-systemd-service host=<robot-ip> ssh_key=<key-path>`
  - [ ] You should get output that looks like this
```
Configuring remote file system to read/write
Pushing system-resource-tracker.service to remote
system-resource-tracker.service             100%  445    63.8KB/s   00:00    
Enabling service
Restarting service
Checking service status
active
Configuring remote file system to read-only
```
- [ ] ssh into robot
  - [ ] Run `cat /data/feature_flags.json` and ensure `enabledPerformanceMetrics` is set to `true`
  - [ ] Run `cat /etc/systemd/system/system-resource-tracker.service` and confirm that the output matches the `system-resource-tracker.service` file in the repo
  - [ ] Run `systemctl status system-resource-tracker | cat` your output should look something like this:
```
● system-resource-tracker.service - Opentrons System Resource Tracker
     Loaded: loaded (/etc/systemd/system/system-resource-tracker.service; enabled; vendor preset: disabled)
     Active: active (running) since Tue 2024-07-09 17:40:20 UTC; 6min ago
   Main PID: 43800 (python3)
      Tasks: 1 (limit: 1761)
     Memory: 9.6M
     CGroup: /system.slice/system-resource-tracker.service
             └─ 43800 python3 -m performance_metrics.system_resource_tracker

Jul 09 17:40:20 9608de python3[43800]: process_filters=('/opt/opentrons*', 'python3*')
Jul 09 17:40:20 9608de python3[43800]: refresh_interval=15.0
Jul 09 17:40:20 9608de python3[43800]: storage_dir=/data/performance_metrics_data
Jul 09 17:40:20 9608de python3[43800]: logging_level=INFO
Jul 09 17:40:20 9608de python3[43800]: 2024-07-09 17:40:20,176 - _metrics_store - setup() - INFO - Setting up metrics store for system_resource_data at /data/performance_metrics_data
Jul 09 17:40:20 9608de /opt/opentrons-robot-server/performance_metrics/system_resource_tracker/__main__.py[43800]: 2024-07-09 17:40:20,174 - __main__ - main() - INFO - Running with the following configuration: 
                                                                                                                   enabled=True
                                                                                                                   process_filters=('/opt/opentrons*', 'python3*')
                                                                                                                   refresh_interval=15.0
                                                                                                                   storage_dir=/data/performance_metrics_data
                                                                                                                   logging_level=INFO
Jul 09 17:40:20 9608de /opt/opentrons-robot-server/performance_metrics/system_resource_tracker/__main__.py[43800]: 2024-07-09 17:40:20,176 - _metrics_store - setup() - INFO - Setting up metrics store for system_resource_data at /data/performance_metrics_data
Jul 09 17:40:20 9608de python3[43800]: 2024-07-09 17:40:20,296 - __main__ - main() - INFO - Starting system resource tracker...
Jul 09 17:40:20 9608de /opt/opentrons-robot-server/performance_metrics/system_resource_tracker/__main__.py[43800]: 2024-07-09 17:40:20,296 - __main__ - main() - INFO - Starting system resource tracker...
Jul 09 17:40:20 9608de systemd[1]: Started Opentrons System Resource Tracker.
```
  - [ ] Run `ls -l /data/performance_metrics_data/` and take note of the size of system_resource_data
  - [ ] Wait like 30 seconds and run ls command again. You should see that `system_resource_data` gets larger
  - [ ] exit ssh
- [ ] Run `make cleanup-remote-server host=<robot-ip> ssh_key=<key-path>`
  - [ ] You should get this output
```
Cleaning up performance metrics on host 10.10.10.151...
Configuring remote file system to read/write
Stopping system-resource-tracker
Disabling system-resource-tracker
Removed /etc/systemd/system/multi-user.target.wants/system-resource-tracker.service.
Reloading daemon
Checking system-resource-tracker status
inactive
Removing service file, data folders, and performance_metrics package
Remounting filesystem as read-only
Unsetting performance metrics feature flag
```

[EXEC-596]: https://opentrons.atlassian.net/browse/EXEC-596?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: performance-metrics/system-resource-tracker.service
Comment (SyntaxColoring) [inline performance-metrics/system-resource-tracker.service:1]: Could you clarify this dev vs. prod split? Is dev meant for `performance-metrics` devs, or robot-software-in-general devs?

File: performance-metrics/system-resource-tracker-dev.service [hunk 1]
```diff
@@ -0,0 +1,19 @@
+[Unit]
+Description=Opentrons System Resource Tracker
+After=opentrons-robot-server.service
+
+[Service]
+Type=notify
+ExecStart=python3 -m performance_metrics.system_resource_tracker
+StateDirectory=system-resource-tracker
+Environment=PYTHONPATH=/opt/opentrons-robot-server
```
Comment (SyntaxColoring) [inline performance-metrics/system-resource-tracker-dev.service:9]: Your call, but it might be worth a comment that this `Environment` value is Flex-specific. The service files elsewhere in our monorepo are OT-2–specific.

File: performance-metrics/Makefile [hunk 2]
```diff
@@ -139,4 +144,60 @@ start-remote-system-resource-tracker:
...
+
+.PHONY: setup-system-resource-tracker-systemd-service
+setup-system-resource-tracker-systemd-service:
+	@echo "Configuring remote file system to read/write"
+	@ssh -i $(ssh_key) root@$(host) "mount -o remount,rw /"
+
+	@echo "Pushing $(FILENAME) to remote"
+	@scp -i $(ssh_key) $(FILENAME) root@$(host):/etc/systemd/system/system-resource-tracker.service
+
+	@echo "Enabling, restarting, and checking systemd service status, then remounting filesystem as read-only"
```
Comment (SyntaxColoring) [inline performance-metrics/Makefile]: Nitpick: to avoid leaving the filesystem writable if something interrupts this sequence, we might want to do it like this:

1. Push the files
2. Run a Bash snippet over SSH that:
    1. Sets an [exit trap](http://redsymbol.net/articles/bash-exit-traps/) that remounts as read-only when the snippet exits
    2. Remounts as writable
    3. Runs `systemctl enable` (and maybe `systemctl daemon-reload`)
4. Run `systemctl restart system-resource-tracker`
    * We have [an existing Makefile function thing](https://github.com/Opentrons/opentrons/blob/0e005dd912900bbd3e2e3c08dcb77fac9c0b7056/scripts/push.mk#L77) for this
6. Run `systemctl is-active system-resource-tracker`, if you need to.

In other words, keep the writable segment as tight as possible, and arrange for it to be undone automatically.

It looks like our other makefiles [already do this](https://github.com/Opentrons/opentrons/blob/0e005dd912900bbd3e2e3c08dcb77fac9c0b7056/scripts/push.mk#L40), except with `&&` and `||` instead of `trap`. That totally works too.

File: performance-metrics/system-resource-tracker-dev.service [hunk 3]
```diff
@@ -0,0 +1,19 @@
+[Unit]
+Description=Opentrons System Resource Tracker
+After=opentrons-robot-server.service
```
Comment (SyntaxColoring) [inline performance-metrics/system-resource-tracker-dev.service:3]: I'm not sure if this exists on the Flex, but on the OT-2, there's an [`opentrons.target`](https://github.com/Opentrons/buildroot/blob/2aefb960558602a87cc0980e6c286fe228ffe88e/board/opentrons/ot2/rootfs-overlay/etc/systemd/system/opentrons.target) that conveniently includes things like [update-server](https://github.com/Opentrons/opentrons/blob/0e005dd912900bbd3e2e3c08dcb77fac9c0b7056/update-server/opentrons-update-server.service#L12) and [Jupyter Notebook](https://github.com/Opentrons/buildroot/blob/2aefb960558602a87cc0980e6c286fe228ffe88e/board/opentrons/ot2/rootfs-overlay/etc/systemd/system/opentrons.target.wants/jupyter-notebook.service). If that exists on Flex, we probably want to use that instead of `opentrons-robot-server.service`.

File: performance-metrics/Makefile [hunk 4]
```diff
@@ -139,4 +144,60 @@ start-remote-system-resource-tracker:
...
+	@ssh -i $(ssh_key) root@$(host) "\
+		systemctl enable system-resource-tracker; \
```
Comment (SyntaxColoring) [inline performance-metrics/Makefile]: Do we need `systemctl daemon-reload` in here somewhere, too?

File: performance-metrics/Makefile [hunk 5]
```diff
@@ -139,4 +144,60 @@ start-remote-system-resource-tracker:
...
+		systemctl restart system-resource-tracker; \
+		systemctl is-active system-resource-tracker; \
```
Comment (SyntaxColoring) [inline performance-metrics/Makefile]: Do you need this `is-active` call? Does `systemctl restart` not already return a failure code if the restart fails?

File: performance-metrics/Makefile [hunk 6]
```diff
@@ -139,4 +144,60 @@ start-remote-system-resource-tracker:
...
+# Default value for the data folder path
+DATA_FOLDER ?= /data/performance_metrics_data
+DEV_DATA_FOLDER ?= /data/performance_metrics_data_dev
+
+cleanup-remote-server:
+	@echo "Cleaning up performance metrics on host $(host)..."
+	
+	@ssh -i $(ssh_key) root@$(host) "\
+		echo "Configuring remote file system to read/write"; \
+		mount -o remount,rw /; \
```
Comment (SyntaxColoring) [inline performance-metrics/Makefile]: For the shell commands in general in this makefile, I think we want to chain most of them with `&&` instead of `;`, so we'll stop as soon as something fails, right? (There might be exceptions to this rule, like the "Checking system-resource-tracker status" step in this cleanup target.)

**Review by SyntaxColoring — APPROVED** [review]
Several minor comments, but this all looks good to me to merge.

Yep, the service files look good to me.

I would add to your test plan to make sure that systemd doesn't log any warnings about the new services. If you misspell a config setting, I think it only logs it as a warning.

File: performance-metrics/system-resource-tracker.service
Comment (DerekMaggio) [reply performance-metrics/system-resource-tracker.service:1]: Robot software in general devs, that way performance can be monitored more granularly. 
But now that I think about it, it might be better to read from a .env file instead. 

Will leave it as is for now

File: performance-metrics/Makefile (hunk 6 above)
Comment (DerekMaggio) [reply performance-metrics/Makefile]: https://github.com/Opentrons/opentrons/pull/15604/commits/99bf43e3c3c7db699e9b4e252ff8f9d109d5a127

File: performance-metrics/Makefile (hunk 5 above)
Comment (DerekMaggio) [reply performance-metrics/Makefile]: https://github.com/Opentrons/opentrons/pull/15604/commits/99bf43e3c3c7db699e9b4e252ff8f9d109d5a127

File: performance-metrics/Makefile (hunk 4 above)
Comment (DerekMaggio) [reply performance-metrics/Makefile]: We only need to reload the daemon if we make a change to any of the service files. 
I did that with https://github.com/Opentrons/opentrons/pull/15604/commits/99bf43e3c3c7db699e9b4e252ff8f9d109d5a127

File: performance-metrics/Makefile (hunk 2 above)
Comment (DerekMaggio) [reply performance-metrics/Makefile]: https://github.com/Opentrons/opentrons/pull/15604/commits/99bf43e3c3c7db699e9b4e252ff8f9d109d5a127
